### ISSUE TITLE
TimeSformer as Recognition Backbone for PoseC3D

### ISSUE THREAD

**Issue by taliegemen** [open-mmlab/mmaction2#1958]
Hello,
Is there a way to use TimeSformer instead of SlowOnlyResnet backbone in PoseC3D. I have made some changes to make the network work but the network does not seem to be learning even if I tried different learning rates. Can you help for this? I am also sending my config file.
Thanks!

```
model = dict(
    type='Recognizer3D',
    backbone=dict(
        type='TimeSformer',
        pretrained= 'https://download.openmmlab.com/mmaction/recognition/timesformer/vit_base_patch16_224.pth', # noqa: E251
	      #pretrained = None,
        num_frames=48,
        img_size=64,
        patch_size=16,
        embed_dims=384,
        in_channels=17,
        dropout_ratio=0.,
        transformer_layers=None,
        attention_type='divided_space_time',
        norm_cfg=dict(type='LN', eps=1e-6)),
    cls_head=dict(type='TimeSformerHead', num_classes=120, in_channels=384),
    # model training and testing settings
    train_cfg=dict(),
    test_cfg=dict(average_clips='prob'))
dataset_type = 'PoseDataset'
ann_file_train = 'data/ntu120_xsub_train.pkl'
ann_file_val = 'data/ntu120_xsub_val.pkl'


left_kp = [1, 3, 5, 7, 9, 11, 13, 15]
right_kp = [2, 4, 6, 8, 10, 12, 14, 16]
train_pipeline = [
    dict(type='UniformSampleFrames', clip_len=48),
    dict(type='PoseDecode'),
    dict(type='PoseCompact', hw_ratio=1., allow_imgpad=True),
    dict(type='Resize', scale=(-1, 64)),
    dict(type='RandomResizedCrop', area_range=(0.5, 1.0)),
    dict(type='Resize', scale=(64, 64), keep_ratio=False),
    dict(type='Flip', flip_ratio=0.5, left_kp=left_kp, right_kp=right_kp),
    dict(
        type='GeneratePoseTarget',
        sigma=0.6,
        use_score=True,
        with_kp=True,
        with_limb=False),
    dict(type='FormatShape', input_format='NCTHW'),
    dict(type='Collect', keys=['imgs', 'label'], meta_keys=[]),
    dict(type='ToTensor', keys=['imgs', 'label'])
]
val_pipeline = [
    dict(type='UniformSampleFrames', clip_len=48, num_clips=1, test_mode=True),
    dict(type='PoseDecode'),
    dict(type='PoseCompact', hw_ratio=1., allow_imgpad=True),
    dict(type='Resize', scale=(-1, 64)),
    dict(type='CenterCrop', crop_size=64),
    dict(
        type='GeneratePoseTarget',
        sigma=0.6,
        use_score=True,
        with_kp=True,
        with_limb=False),
    dict(type='FormatShape', input_format='NCTHW'),
    dict(type='Collect', keys=['imgs', 'label'], meta_keys=[]),
    dict(type='ToTensor', keys=['imgs'])
]
test_pipeline = [
    dict(
        type='UniformSampleFrames', clip_len=48, num_clips=10, test_mode=True),
    dict(type='PoseDecode'),
    dict(type='PoseCompact', hw_ratio=1., allow_imgpad=True),
    dict(type='Resize', scale=(-1, 64)),
    dict(type='CenterCrop', crop_size=64),
    dict(
        type='GeneratePoseTarget',
        sigma=0.6,
        use_score=True,
        with_kp=True,
        with_limb=False,
        double=True,
        left_kp=left_kp,
        right_kp=right_kp),
    dict(type='FormatShape', input_format='NCTHW'),
    dict(type='Collect', keys=['imgs', 'label'], meta_keys=[]),
    dict(type='ToTensor', keys=['imgs'])
]
data = dict(
    videos_per_gpu=8, #16
    workers_per_gpu=2, #2 were older
    test_dataloader=dict(videos_per_gpu=1),
    train=dict(
        type=dataset_type,
        ann_file=ann_file_train,
        data_prefix='',
        class_prob={i: 1 + int(i >= 60)
                    for i in range(120)},
        pipeline=train_pipeline),
    val=dict(
        type=dataset_type,
        ann_file=ann_file_val,
        data_prefix='',
        pipeline=val_pipeline),
    test=dict(
        type=dataset_type,
        ann_file=ann_file_val,
        data_prefix='',
        pipeline=test_pipeline))
# optimizer
optimizer = dict(
    type='SGD', lr=1e-4, momentum=0.9,
    weight_decay=0.0003)  # this lr is used for 8 gpus * 16 vid per was 0.2
optimizer_config = dict(grad_clip=dict(max_norm=40, norm_type=2))
# learning policy
lr_config = dict(policy='CosineAnnealing', by_epoch=False, min_lr=0)
total_epochs = 250 # was 240
checkpoint_config = dict(interval=1)
workflow = [('train', 10)]
evaluation = dict(
    interval=10,
    metrics=['top_k_accuracy', 'mean_class_accuracy'],
    topk=(1, 5))
log_config = dict(
    interval=20, hooks=[
        dict(type='TextLoggerHook'),
    ])
dist_params = dict(backend='nccl')
log_level = 'INFO'
work_dir = './work_dirs/posec3d/slowonly_r50_u48_240e_ntu120_xsub_keypoint'
load_from = None
#load_from = 'data/slowonly_r50_u48_240e_ntu120_xsub_keypoint.pth'
#my edit on line 131
resume_from = None
find_unused_parameters = False
gpu_ids = [0] #added by me for inference

```

**Comment by taliegemen** [comment]
Ok, I think it works. It was just a problem with learning rate and total epochs in my opinion. top5 accuracy is around 0.5 at this moment of the training, I will keep you updated if I get any problems.

**Comment by ahmed-nady** [comment]
@taliegemen what is the result of using TimeSformer as the backbone? and on which datasets?

**Comment by taliegemen** [comment]
I trained it with NTU 120 dataset, but scores are rather low (0.73 top1). Maybe something is causing overfitting but I did not be able to fix it by changing dropout or weight decay.

**Comment by sin300** [comment]
@taliegemen Hello, I am also bothered by the same problem, could you tell me how you changed the learning rate and total epochs to solve this problem?Thanks alot!

**Comment by taliegemen** [comment]
Hello @sin300,
For some reason, scores will always be low but I manage to get around 0.75 by adjusting the patch size if I remember correctly. In normal means, transformer often uses 16 as patch size (there are some papers suggest that after 16 you start to get lower accuracy results) but in this case I would suggest you to use 32 as your patch size as it will offer better accuracy. I will also recommed you to ditch SGD for this one and instead opt for AdamW optimizer, as it performs better for timesformer, and lower the decay (maybe 0.05 decay or 0.1 decay) since timesformer seems not be working well when there is too much decay unlike convolutional alternatives.  After those you will just be needing to get the correct learning rate for your dataset, something like 0.001 or 0.01 will work well here.
Let me know if you have any more questions, and ofc you can cite our paper if those helps, when it is published :P
All the best,

**Comment by sin300** [comment]
@taliegemen Thank you very much for your rapid and detailed response!As your suggested,i uses 32 as patch size and  AdamW optimizer with 0.1 decay.But they don't work on my custom dataset,which can be trained well on PoseC3d model.Here is my config file.

```python
num_frames = 32  
img_size = 64  
model = dict(  
    type='Recognizer3D',  
    backbone=dict(  
        type='TimeSformer',  
        num_frames=num_frames,  
        img_size=img_size,  
        patch_size=32,  
        embed_dims=384, 
        in_channels=17,  
        dropout_ratio=0.,  
        transformer_layers=None,  
        attention_type='divided_space_time',  
        norm_cfg=dict(type='LN', eps=1e-6)),  
    cls_head=dict(  
        type='TimeSformerHead',  
        num_classes=8,  
        in_channels=384, ))  
  
dataset_type = 'PoseDataset'  
ann_file = 'volleyball_hrnet.pkl'  
left_kp = [1, 3, 5, 7, 9, 11, 13, 15]  
right_kp = [2, 4, 6, 8, 10, 12, 14, 16]  
  
train_pipeline = [  
    dict(type='UniformSampleFrames', clip_len=num_frames),  
    dict(type='PoseDecode'),  
    dict(type='PoseCompact', hw_ratio=1., allow_imgpad=True),  
    dict(type='RandomResizedCrop', area_range=(0.7, 1.0)),  
    dict(type='Resize', scale=(img_size, img_size), keep_ratio=False),  
    dict(type='GeneratePoseTarget', with_kp=True, with_limb=False),  #  
    dict(type='FormatShape', input_format='NCTHW_Heatmap'),  
    dict(type='Collect', keys=['imgs', 'label'], meta_keys=[]),  
    dict(type='ToTensor', keys=['imgs', 'label'])  
]  
val_pipeline = [  
    dict(type='UniformSampleFrames', clip_len=num_frames, num_clips=1),  
    dict(type='PoseDecode'),  
    dict(type='PoseCompact', hw_ratio=1., allow_imgpad=True),  
    dict(type='Resize', scale=(img_size, img_size), keep_ratio=False),  
    dict(type='GeneratePoseTarget', with_kp=True, with_limb=False),  #  
    dict(type='FormatShape', input_format='NCTHW_Heatmap'),  
    dict(type='Collect', keys=['imgs', 'label'], meta_keys=[]),  
    dict(type='ToTensor', keys=['imgs'])  
]  
test_pipeline = [  
    dict(type='UniformSampleFrames', clip_len=num_frames, num_clips=1),  
    dict(type='PoseDecode'),  
    dict(type='PoseCompact', hw_ratio=1., allow_imgpad=True),  
    dict(type='Resize', scale=(img_size, img_size), keep_ratio=False),  
    dict(type='GeneratePoseTarget', with_kp=True, with_limb=False),  #  
    dict(type='FormatShape', input_format='NCTHW_Heatmap'),  
    dict(type='Collect', keys=['imgs', 'label'], meta_keys=[]),  
    dict(type='ToTensor', keys=['imgs'])  
]  
data = dict(  
    videos_per_gpu=8,  
    workers_per_gpu=2,  
    val_dataloader=dict(videos_per_gpu=1),  
    test_dataloader=dict(videos_per_gpu=1),  
    train=dict(  
        type='RepeatDataset',  
        times=10,  
        dataset=dict(type=dataset_type, ann_file=ann_file, split='train', pipeline=train_pipeline)),  
    val=dict(type=dataset_type, ann_file=ann_file, split='test', pipeline=val_pipeline),  
    test=dict(type=dataset_type, ann_file=ann_file, split='test', pipeline=test_pipeline))  

optimizer = dict(type='AdamW', lr=0.001, weight_decay=0.1,  
                 paramwise_cfg=dict(  
                     custom_keys={'.backbone.cls_token': dict(decay_mult=0.0),  
                                  '.backbone.pos_embed': dict(decay_mult=0.0),  
                                  '.backbone.time_embed': dict(decay_mult=0.0)  
                                  }))  #  
  
optimizer_config = dict(grad_clip=dict(max_norm=40, norm_type=2))  
# learning policy  
lr_config = dict(policy='CosineAnnealing', by_epoch=False, min_lr=0)  
total_epochs = 10  
checkpoint_config = dict(interval=1)  
evaluation = dict(interval=1, metrics=['top_k_accuracy', 'mean_class_accuracy'], topk=1)  
log_config = dict(interval=20, hooks=[dict(type='TextLoggerHook')])  

```
I wonder know how you change embed_dims and img_size after using 32 as patch size.Maybe the values of these two parameters are not reasonable in my config.

Hope your paper can be successfully published and i'm very glad to cite it if it helps.Best wishes!

**Comment by taliegemen** [comment]
@sin300 my pleasure,
From the settings I would try to change the following:
`embed_dims=768
videos_per_gpu=16
in_channels=768
dropout=0.1
`
I think dropout and embed dims matter much, maybe also batch size, I would just try to increase them at this point (if i remember correctly you will be needing to increase the input channels to increase embed dims). I will also take a look at my code this evening (if I don't forget it :S) and try to see what is wrong with this one.
